features/mbedtls/targets/TARGET_Silicon_Labs/crypto_aes.c

```diff
@@ -0,0 +1,545 @@
+            CORE_ENTER_CRITICAL();
+            CRYPTO_DataWriteUnaligned(&device->DATA0XOR, (uint8_t *)(&input[processed]));
+            device->CMD = CRYPTO_CMD_INSTR_AESENC;
+            CRYPTO_DataReadUnaligned(&device->DATA0, (uint8_t *)(&output[processed]));
+            CORE_EXIT_CRITICAL();
+        } else {
+            /* Decrypt input block, XOR IV to decrypted text, set ciphertext as next IV */
+            CORE_ENTER_CRITICAL();
+            CRYPTO_DataWriteUnaligned(&device->DATA0, (uint8_t *)(&input[processed]));
+            CRYPTO_EXECUTE_4( device,
+                              CRYPTO_CMD_INSTR_DATA0TODATA1,
+                              CRYPTO_CMD_INSTR_AESDEC,
+                              CRYPTO_CMD_INSTR_DATA2TODATA0XOR,
+                              CRYPTO_CMD_INSTR_DATA1TODATA2);
+            CRYPTO_DataReadUnaligned(&device->DATA0, (uint8_t *)(&output[processed]));
+            CORE_EXIT_CRITICAL();
+        }
+        processed += 16;
+    }
+
+    if ( processed >= 16 ) {
+        if ( mode == MBEDTLS_AES_ENCRYPT ) {
+            memcpy(iv, &output[processed-16], 16);
+        } else {
+            CORE_ENTER_CRITICAL();
+            CRYPTO_DataReadUnaligned(&device->DATA2, (uint8_t *)(iv));
+            CORE_EXIT_CRITICAL();
+        }
+    }
+
+    crypto_management_release(device);
+
+    return ret;
+}
+#endif /* MBEDTLS_CIPHER_MODE_CBC */
+
+#if defined(MBEDTLS_CIPHER_MODE_CFB)
+/*
+ * AES-CFB128 buffer encryption/decryption
+ */
+int mbedtls_aes_crypt_cfb128( mbedtls_aes_context *ctx,
+                              int mode,
+                              size_t length,
+                              size_t *iv_off,
+                              unsigned char iv[16],
+                              const unsigned char *input,
+                              unsigned char *output )
+{
+    size_t n = iv_off ? *iv_off : 0;
+    size_t processed = 0;
+    int ret = 0;
+    CORE_DECLARE_IRQ_STATE;
+
+    if( ctx == NULL || input == NULL || output == NULL || iv == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    if ( ctx->keybits != 128UL && ctx->keybits != 256UL) {
+        return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+    }
+
+    while ( processed < length ) {
+        if ( n > 0 ) {
+            /* start by filling up the IV */
+            if( mode == MBEDTLS_AES_ENCRYPT ) {
+                iv[n] = output[processed] = (unsigned char)( iv[n] ^ input[processed] );
+            } else {
+                int c = input[processed];
+                output[processed] = (unsigned char)( c ^ iv[n] );
+                iv[n] = (unsigned char) c;
+            }
+            n = ( n + 1 ) & 0x0F;
+            processed++;
+            continue;
+        } else {
+            /* process one ore more blocks of data */
+            CRYPTO_TypeDef *device = crypto_management_acquire();
+            device->WAC = 0;
+            device->CTRL = 0;
+
+            CORE_ENTER_CRITICAL();
```

Comment:
The resource we are protecting with critical sections is writing to or reading from 128-bit, 256-bit or 512-bit registers in our accelerator. Since that operation is done with actionable reads/writes from/to a single 32-bit register, they have to be protected against preemption.

The other registers (such as WAC and CTRL) are regular registers, which we get control over by acquiring the device. In the case of preemption, these values will be backed up and put back by the preempting code, so their use is safe.